Add types to events and use these to check the parser

trarbr opened the pull request:
This was a small thing I wanted to do that turned out to be a pretty big diff - sorry about that 😄 

The main purpose is to make it easier to figure out what the possible values of fields on events can be - what are the names of the different maps, game modes, weapons, etc. All the possible values for these "constant types" are now defined in `CsgoStats.Types`.

In addition, I also wanted to make the parser be more strict, in order to guarantee that these are in fact the only values that will be parsed out - if it tries to parse a game map or a weapon that is unknown, the parser should skip the event, instead of sending it downstream - in order to avoid missing logic and/or assets for those cases. 

In some cases, re-using the type definitions in the parser makes the parser module a lot smaller - e.g. `CsgoStats.Logs.Parser.Weapon`. This is the case when the strings directly match the atom names. I can simply map over the list of possible atoms for a weapon  and translate them to parser combinators with `NimbleParsec.string/1`. And then I can safely cast the weapon's string name to an atom with `String.to_existing_atom/1` since I know the atom has been defined. Now, to add support for a new weapon type I just have to add the weapon to the list in `CsgoStats.Types` and the parser definition updates automatically. 

When the strings don't match the atom names, the code is a bit more complex 🤢 An example is `CsgoStats.Logs.Parser.Item`. In these cases I use another approach. I still hardcode the parser combinators for strings, but then define translation functions by using the list of possible values defined in `CsgoStats.Types`. When done this way, ElixirLS can generate a warning if my definition is not complete. E.g. if I define the `cast_item` function, but forget a match for `:c4`, I get a compile error like this: 

<img width="790" alt="Screenshot 2020-10-30 at 07 59 15" src="https://user-images.githubusercontent.com/6442970/97670143-e7641300-1a85-11eb-9b92-2ed6a8c9a807.png">

So even though the code is a little unusual, I like the fact that it can help make sure we match all the possible cases.

Commit:
Add types to events and use these to check the parser

nicoeg committed:
Point to game_map instead of map

nicoeg reviewed (approved):
Wow this is nice :ok_hand: 

I just added a commit that fixes an error because match.ex was still pointing to `map` instead of `game_map` on line 95. Hope it's ok.

Commit:
Rename Match map field to game_map

trarbr commented:
> Wow this is nice 👌
> 
> I just added a commit that fixes an error because match.ex was still pointing to `map` instead of `game_map` on line 95. Hope it's ok.

Thanks, nice catch. I just renamed the `map` field on Match to `game_map` as well.
